✨ Feat: #12 Level 4. UIScrollView를 추가하여 목차를 구성

amazingDove367 opened the pull request:
# Level 4: UIScrollView를 추가하여 목차를 구성
### 구현내용
1. 스크롤 속성
   - `UIScrollView` 를 추가하여 스크롤할 수 있도록 구현한 후 목차(Chapters)를 구성
      - 책 제목과 시리즈 순서는 화면 상단에 고정
      - 책 정보(책 표지, 책 제목, 저자, 출간일, 페이지수) 영역부터 목차(Chapters)까지 스크롤 가능하도록 구현
2. 목차 속성
   - `UIScrollView` 안에 `UIStackView`를 추가하고 해당 `UIStackView`안에 `UILabel` 추가하여 구현
       - 전체적인 포함 관계: `UIScrollView` 안에 `UIStackView` 안에 `UILabel`들어 있는 구조

amazingDove367 committed:
✨ Feat: #12 Level 4. UIScrollView를 추가하여 목차를 구성
1. 스크롤 속성
   - UIScrollView 를 추가하여 스크롤할 수 있도록 구현한 후 목차(Chapters)를 구성
     - 책 제목과 시리즈 순서는 화면 상단에 고정
     - 책 정보(책 표지, 책 제목, 저자, 출간일, 페이지수) 영역부터 목차(Chapters)까지 스크롤 가능하도록 구현
2. 목차 속성
   - UIScrollView 안에 UIStackView를 추가하고 해당 UIStackView안에 UILabel 추가하여 구현
   - 전체적인 포함 관계: UIScrollView 안에 UIStackView 안에 UILabel들어 있는 구조

amazingDove367 committed:
🔥 Remove: #14 Level 4. Preview 코드 제거
잘못 올라간 Preview 코드 삭제

amazingDove367 committed:
🔥 Remove: #15 Level 4. demo chapter 코드 제거
demo 값으로 넣어놓은 chapter 값 삭제

amazingDove367 committed:
🔥 Remove: #16 Level 4. HomeView 파일 Autolayout constraints 하드코드 제거
HomeView.swift 파일 Autolayout constraints 중, 하드코드 제거 / Constants type  상수선언

dawnie-g commented on HarryPotterSeriesHakyung/HarryPotterSeriesHakyung/View/HomeView.swift:
```diff
@@ -94,6 +119,15 @@ final class HomeView: UIView {
         numberButton.backgroundColor = Constants.Components.buttonColor
         numberButton.layer.cornerRadius = Constants.Components.buttonLength / 2
+        
+        scrollview.isScrollEnabled = true
```
이 속성은 default 값이 true여서 뺴주셔도 괜찮아요

amazingDove367 replied:
default군요!! 감사합니다 :)

dawnie-g commented on HarryPotterSeriesHakyung/HarryPotterSeriesHakyung/View/HomeView.swift:
```diff
@@ -94,6 +119,15 @@ final class HomeView: UIView {
+        
+        scrollview.isScrollEnabled = true
+        scrollview.alwaysBounceVertical = true
+        scrollview.showsVerticalScrollIndicator = true
```
요구사항에서 인디케이터 안보이게 구현 해야한다고 되어있어요!

amazingDove367 replied:
엇 놓쳤네요! 수정하겠습니다!

dawnie-g commented on HarryPotterSeriesHakyung/HarryPotterSeriesHakyung/View/ChapterView.swift:
```diff
@@ -0,0 +1,90 @@
+            vStackView.addArrangedSubview(chapter)
+            return chapter
+        }
+        vStackView.layoutIfNeeded()
```
`layoutIfNeeded()`를 호출하지 않으면 UI 업데이트가 안되나요?

amazingDove367 replied:
`layoutIfNeeded` 이거 빼보니 작동 잘 합니다!

어제 구현하면서 동적으로 데이터가 바뀔 경우에, Chapters의 title 컴포넌트들이 생성되지 않는 이슈가 있었는데요, 
`vStackView`가 업데이트가 안됐구나! 해서 넣었던 코드인데, 이게 문제가 아니었어요 ㅋ

핵심은 `vStackView.addArrangedSubview(chapter)`로 `vStackView`에 넣어주지 않고 있던 것이 원인이었습니다!

아래 `configSubview()` 부분에서 뷰를 로드할때 그려주고 있으니까 업데이트 된 후에도  `vStackView`를 업데이트 해주면 
 Chapters의 title 컴포넌트들 갯수에 맞게 생성되고 업데이트 된다고 생각했나봐요 ㅎ

```swift
private func configSubview() {
        ...
        chapterContents
            .forEach({ vStackView.addArrangedSubview($0) })
    }
```

dawnie-g replied:
오... `addArrangedSubview(_:)`로 넣어주지 않고 `addSubview(_:)`를 사용하게 되면 UI 업데이트를 수동으로 해줘야하는 거군요? 몰랐던 사실인데 감사합니다! 😘

dawnie-g commented on HarryPotterSeriesHakyung/HarryPotterSeriesHakyung/View/HomeView.swift:
```diff
@@ -65,19 +71,38 @@ final class HomeView: UIView {
+        [titleLable, numberButton]
+            .forEach { contentView.addSubview($0) }
+        
+        [contentSubView, vStackView]
```
`contentSubview`는 제약도 걸려있고 추가도 되어 있는데 어떤 서뷰뷰를 갖고 있나요?

dawnie-g reviewed (approved):
새벽까지 구현 해내주시느라 고생 많으셨어요..!
오늘도 화이팅💪💪

meowbutlerdev commented on HarryPotterSeriesHakyung/HarryPotterSeriesHakyung/View/HomeView.swift:
```diff
@@ -102,21 +136,43 @@ final class HomeView: UIView {
         infoView.snp.makeConstraints {
-            $0.top.equalTo(numberButton.snp.bottom).offset(Constants.Spacing.spacing30)
-            $0.horizontalEdges.equalToSuperview()
+            $0.top.equalTo(vStackView.snp.top).offset(Constants.Spacing.spacing30)
```
여기서 제약조건 충돌 발생하네요.
스택뷰 내부 조건이랑 충돌되는 것 같습니다.

meowbutlerdev reviewed (approved):
그리고 추가로 `dedicationView` 제약 조건도 충돌이 발생하고 있어요.
(하경님이 빌드하실 땐 안 날 수도 있겠지만..)

또, 스크롤이 제대로 되지 않고 있습니다.
스크롤하면 다시 제자리로 돌아가네요.
이것도 저만 그런걸까요?

amazingDove367 commented:
> 또, 스크롤이 제대로 되지 않고 있습니다.
> 스크롤하면 다시 제자리로 돌아가네요.
> 이것도 저만 그런걸까요?

맞아요.. 이거 고치다가 Chapters title들의 크기를 못읽는건지, 읽는데 제약사항 업데이트가 안되는 건지, 제약사항이 잘못된건지.. 
원인을 다 알아보지 못하고 일단은 종료했었습니다
일단은 버그로 빼놓겠습니다!



> 그리고 추가로 `dedicationView` 제약 조건도 충돌이 발생하고 있어요.

네 저도 노랑색 경고 뜨고 있어요.. 이것도 수정사항으로 빼놓겠습니다!

amazingDove367 committed:
♻️ Refactor: # 19 Level 4: 코드리뷰 대응
1. scrollview.isScrollEnabled = true는 default속성이므로 제거
2. layoutIfNeeded()를 호출하지 않으면 UI 업데이트가 안되는지
3. 요구사항에서 showsVerticalScrollIndicator, alwaysBounceVertical 보이지 않는다고 명시되어있음
